### DISCUSSION TITLE
Redirect blob/tree urls when a branch is renamed

### DISCUSSION THREAD

**Discussion by eps1lon** [community/community#23213] [New to GitHub]
<p>If we want to rename our default branch from <code>master</code> to <code>main</code> then we need to make sure every link that pointed to <code>master</code> is updated as well.</p>
<p>This is sometimes not possible if the published document is locked. I realize that we just “shouldn’t have linked to a branch because the file gets removed” but for evergreen files such as “README” or “CONTRIBUTING” this would be a bit unfortunate.</p>
<p>If GitHub would redirect links to tree and blob (e.g. <a href="https://github.com/eps1lon/dom-accessibility-api/blob/f4293f4d373944b583797a960fda958873bb28b9/README.md" rel="nofollow noopener">https://github.com/eps1lon/dom-accessibility-api/blob/f4293f4d373944b583797a960fda958873bb28b9/README.md</a>) for renamed branches we wouldn’t have this issue.</p>

**Comment by yamiacat** [comment]
<p>Hey eps1lon! :wave:</p>
<p>Thanks for this feedback! We’re always working to improve GitHub and the GitHub Support Community, and we consider every suggestion we receive.</p>
<p>Would you mind <a href="https://support.github.com/contact/feedback">submitting this through our official product feedback form</a> so that our product team can track your request?</p>

**Comment by eps1lon** [comment]
<p>What should I fill in under “What part of GitHub is your feedback about?”?</p>

**Comment by yamiacat** [comment]
<p>I think this would come under “other”! That means it goes to the main Product Team, rather than one of the more specialized sub=teams…</p>

**Comment by dwmarshall** [comment]
<p>A git repository doesn’t know how branches have been renamed, so I don’t know how this would work. I did suggest a few weeks ago that there should be a synthetic branch name such as <code>_default</code> that would resolve to whatever a repository’s default branch is. However, this may be premature optimization. How often are we going to paradigmatically change the default branch name? After some discussion at my workplace, <code>main</code> is the new default.</p>
<p>Perhaps the suggestion should be that if a URL references a branch that does not exist, it should be redirected to the current default branch with an advisory.</p>

**Comment by eps1lon** [comment]
<blockquote>
<p>A git repository doesn’t know how branches have been renamed</p>
</blockquote>
<p>Sure, github can know it though if it allows branch renaming within its UI.</p>
<p>A magic name for default branches wouldn’t help in our case since we go through cycles where main development happens on main or alpha branches. Having docs link to alpha branches would not be ideal.</p>

**Comment by briandealwis** [comment]
<p>One of my work colleagues discovered that you can use HEAD to reference through the project’s default branch.</p>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/eps1lon/dom-accessibility-api/blob/HEAD/README.md" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/eps1lon/dom-accessibility-api/blob/HEAD/README.md" target="_blank" rel="nofollow noopener">eps1lon/dom-accessibility-api/blob/HEAD/README.md</a></h4>
<pre><code class="lang-md"># dom-accessibility-api

[![npm version](https://badge.fury.io/js/dom-accessibility-api.svg)](https://badge.fury.io/js/dom-accessibility-api)
[![Build Status](https://dev.azure.com/silbermannsebastian/dom-accessibility-api/_apis/build/status/eps1lon.dom-accessibility-api?branchName=main)](https://dev.azure.com/silbermannsebastian/dom-accessibility-api/_build/latest?definitionId=6&amp;branchName=main)
![Azure DevOps coverage](https://img.shields.io/azure-devops/coverage/silbermannsebastian/dom-accessibility-api/6)

Computes the accessible name or description of a given DOM Element.
https://w3c.github.io/accname/ implemented in JavaScript for testing.

```bash
$ yarn add dom-accessibility-api
```

```js
import {
	computeAccessibleName,
	computeAccessibleDescription,
} from "dom-accessibility-api";
```

</code></pre>

  This file has been truncated. <a href="https://github.com/eps1lon/dom-accessibility-api/blob/HEAD/README.md" target="_blank" rel="nofollow noopener">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by eps1lon** [comment]
<p>Thanks! That works for some use cases. If HEAD always points to a “stable” branch or at least candidates for stable then all is fine. But sometimes we go through cycles where development happens on an unstable branch I wouldn’t want links to point to unstable work.</p>
<p>Adding redirects for branch renames comes down to “don’t break the web”</p>

**Comment by eps1lon** [comment, marked as answer]
<p><a href="https://github.blog/changelog/2020-07-17-links-to-deleted-branches-now-redirect-to-the-default-branch/" rel="noopener nofollow ugc">GitHub now redirects broken view links to the default branch</a>.</p>
